sparkmagic/sparkmagic/livyclientlib/livysession.py

```diff
@@ -17,38 +17,56 @@
 
 class _HeartbeatThread(threading.Thread):
     def __init__(self, livy_session, refresh_seconds, retry_seconds, run_at_most=None):
+        """
+        Parameters
+        ----------
+        livy_session : LivySession
+        refresh_seconds: int
+            The seconds to sleep between refreshing the Livy session status and info
+        retry_seconds: int
+            The seconds to sleep before retrying on a failed refresh
+        run_at_most: int, optional
+            The max number of loops to execute before ending this thread
+        """
         super(_HeartbeatThread, self).__init__()
-        
+
         self.livy_session = livy_session
         self.refresh_seconds = refresh_seconds
         self.retry_seconds = retry_seconds
+        if run_at_most is None:
+            # a billion iterations should be sufficient
+            run_at_most = int(1e9)
         self.run_at_most = run_at_most
 
     def run(self):
-        i = 0
-        if self.livy_session is not None:
-            self.livy_session.logger.info(u'Starting heartbeat for session {}'.format(self.livy_session.id))
-        else:
-            self.livy_session.logger.info(u'Will not start heartbeat because session is none')
```

Comment:
Hi @itamarst. This is the offending line. On line 29, we are testing for `self.livy_session` being `None`. If the `else` bit of the clause gets executed, then `self.livy_session` is guaranteed to be `None` and so we will get a stack trace saying roughly "None type has no attribute 'logger'", which is a horribly cryptic and confusing error message. This PR fixes that issue, as well as adding some docs, giving better variable names and simplifying some looping logic.

Reply:
Ah hah, makes sense.